feat: editorial on Tree Painting (#3455)
cpinitiative/usaco-guide: solutions/gold/cf-1187E.mdx

the wrong initial node with an inoptimal ending cost.

We can calculate these initial values with another DFS search before the one
calculating the values for $dp_2$. That way we can remove the $sz[u] - sz[c]$
from the recurrence formula for $dp_2$. To calculate these initial values, we
notice that every time we go down by one depth, we add size of all the siblings
and parent combined two times, one for the parent, and once for the current
node. We also add the cost of all the ancestor's siblings and parent combined
once more. Let's say $p_2[u]$ is the initial cost for a leaf node $u$, and
$p_1[u]$ is the cost of the ancestor's siblings and parent combined till $u$.
Then we can calculate $p_1[u]$ and $p_2[u]$ using the following recurrence:

$$
p_1[u] = p_1[p] + sz[p] - sz[u] \newline
p_2[u] = p_2[p] + p_1[u] + sz[p] - sz[u]
$$

## Implementation

**Time Complexity:** $\mathcal{O}(V+E)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

vector<vector<ll>> g;
vector<ll> dp1, dp2;
vector<ll> p1, p2;
vector<ll> sz;

// Calculate sz and dp1
void dfs1(ll u, ll p) {
	sz[u] = 1;

> 	for (auto &v: g[u]) {

Reviewer: prefer if you didn't use auto

> 		if (v == p) continue;

Reviewer: braces

Reviewer: in lotta other places asw

		dfs1(v, u);
		sz[u] += sz[v];
		dp1[u] += dp1[v];
	}
	dp1[u] += sz[u];
}

// Calculate cost incorporated for initial value of dp2
void dfs2(ll u, ll p) {
	if (p != -1) {
		p1[u] = p1[p] + sz[p] - sz[u];
		p2[u] = p2[p] + p1[u] + sz[p] - sz[u];
	}
	for (auto &v: g[u]) {
		if (v == p) continue;
		dfs2(v, u);
	}
}

// Calculate dp2
void dfs3(ll u, ll p) {
	dp2[u] = (g[u].size() == 1 ? p2[u]+1 : 0);
	ll tot{0};
	for (auto &v: g[u]) {
		if (v == p) continue;
		tot += dp1[v];
	}
	for (auto &v: g[u]) {
		if (v == p) continue;
		dfs3(v, u);
		if (dp2[v] + tot - dp1[v] > dp2[u]) {
			dp2[u] = dp2[v] + tot - dp1[v];
		}
	}
}

int main() {

> 	ios_base::sync_with_stdio(false);

Reviewer: remove if runs w/o

> 	ll n;

Reviewer: idt the nodes have to be `ll`?

	cin >> n;

	g.resize(n); dp1.resize(n); dp2.resize(n); sz.resize(n);
	p1.resize(n); p2.resize(n);
	for (ll i = 1; i < n; i++) {
		ll x, y;
		cin >> x >> y;
		g[--x].push_back(--y);
		g[y].push_back(x);
	}

	// Run first DFS to calculate sz and dp1
	dfs1(0, -1);

	// Run second DFS to calculate p1 and p2
	dfs2(0, -1);

	// Run third DFS to calculate dp2
	dfs3(0, -1);

	// Find maximum dp1/dp2 among all nodes

> 	ll ans{0};

Reviewer: ```suggestion
	ll ans = 0;
```

---
id: cf-1187E
source: CF
title: Tree Painting
author: Yash Singh
---

[Official Editorial](https://codeforces.com/blog/entry/68111)

## Explanation

> We notice that choosing a vertice splits the current connected component into

Reviewer: > vertice

i think it's vertex

Reviewer: also this first sentence,,, seems kinda obvious?

the childrens' subtrees and the rest of the tree excluding the current subtree.
Another observation we can make is that choosing a leaf node is more optimal
than choosing anything in the middle because that way we have the entire size of
the rest of the tree as a part of the leftover connected component. This hints
for a DP on Trees solution, we choose one leaf node, work our way up and work

> our way down for each of the siblings. For example, for the following tree:

Reviewer: hb uh
> This motivates a tree DP solution where we [do stuff]

what does "work up"/"work down" mean
also siblings of what? the leaf node?

Author: ya i worded that really bad lemme see if i can make it better

Author: I changed the wording a bit and the pic for the tree, could you check now?

![Tree Visualization](cf-1187E/tree.png)

If we select the 4th node as our starting point, then the process would look
like the following:

> 1. Work our way up to 3

Reviewer: again, what does "work our way up" mean

Author: Work our way up means to start from the leaf node and go through each of the ancestors.

Reviewer: and...what are you doing when going through the ancestors?

later on which would have prefered a higher depth over a higher cost. This means
that we need to incorporate the $sz[u] - sz[c]$ for every ancestor node when
initializing the $dp_2$ value for leaf nodes. That way we, it will be
incorporated into the cost beforehand and we don't need to worry about choosing
the wrong initial node with an inoptimal ending cost.

We can calculate these initial values with another DFS search before the one
calculating the values for $dp_2$. That way we can remove the $sz[u] - sz[c]$
from the recurrence formula for $dp_2$. To calculate these initial values, we
notice that every time we go down by one depth, we add size of all the siblings
and parent combined two times, one for the parent, and once for the current
node. We also add the cost of all the ancestor's siblings and parent combined
once more. Let's say $p_2[u]$ is the initial cost for a leaf node $u$, and
$p_1[u]$ is the cost of the ancestor's siblings and parent combined till $u$.
Then we can calculate $p_1[u]$ and $p_2[u]$ using the following recurrence:

$$
p_1[u] = p_1[p] + sz[p] - sz[u] \newline
p_2[u] = p_2[p] + p_1[u] + sz[p] - sz[u]
$$

## Implementation

> **Time Complexity:** $\mathcal{O}(V+E)$

Reviewer: put these in terms of the problem's constraints

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

vector<vector<int>> g;
vector<ll> dp1, dp2;
vector<ll> p1, p2;
vector<ll> sz;

// Calculate sz and dp1
void dfs1(int u, int p) {
	sz[u] = 1;
	for (int &v: g[u]) {
		if (v == p) {
			continue;
      dfs1(v, u);

> 		}

Reviewer: sus

Author: o lol i was editing in vscode it prob reset the spaces

parent combined once more. Let's say $p_2[u]$ is the initial cost for a leaf
node $u$, and $p_1[u]$ is the cost of the ancestor's siblings and parent
combined till $u$. Then we can calculate $p_1[u]$ and $p_2[u]$ using the
following recurrence:

$$
p_1[u] = p_1[p] + sz[p] - sz[u] \newline
p_2[u] = p_2[p] + p_1[u] + sz[p] - sz[u]
$$

## Implementation

**Time Complexity:** $\mathcal{O}(N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

vector<vector<int>> g;
vector<ll> dp1, dp2;
vector<ll> p1, p2;
vector<ll> sz;

// Calculate sz and dp1
void dfs1(int u, int p) {
	sz[u] = 1;
	for (int &v: g[u]) {
		if (v == p) {
			continue;
		}
		dfs1(v, u);
		sz[u] += sz[v];
		dp1[u] += dp1[v];
	}
	dp1[u] += sz[u];
}

// Calculate cost incorporated for initial value of dp2
void dfs2(int u, int p) {
	if (p != -1) {
		p1[u] = p1[p] + sz[p] - sz[u];
		p2[u] = p2[p] + p1[u] + sz[p] - sz[u];
	}
	for (int &v: g[u]) {
		if (v == p) {
			continue;
		}
		dfs2(v, u);
	}
}

// Calculate dp2
void dfs3(int u, int p) {
	dp2[u] = (g[u].size() == 1 ? p2[u]+1 : 0);
	ll tot = 0;

> 	for (int &v: g[u]) {

Reviewer 2: theres no reason to pass in ref for all these for loops

just a personal preference but it looks ugly lol

## Explanation

> An observation we can make is that choosing a leaf node is more optimal than

Reviewer: ```suggestion
Notice that choosing a leaf node is more optimal than
```
